### PR TITLE
Extend alias analysis to list interesting paramters of each procedure

### PR DESCRIPTION
**Pull request by CaviarChen**
This is part of #31

### CODE REVIEW

File: src/AliasAnalysis.hs [hunk 1]
```diff
@@ -20,14 +20,16 @@ import           Data.List     as List
...
+import           Flow          ((|>))
 import           Options       (LogSelection (Analysis))
 import           Util
 
 
-----------------------------------------------------------------
---                 Proc Level Aliasing Analysis
-----------------------------------------------------------------
--- PROC LEVEL ALIAS ANALYSIS
+-- for local use only
```
Comment (pschachte) [inline src/AliasAnalysis.hs:28]: Please document this type

File: src/AliasAnalysis.hs [hunk 2]
```diff
@@ -51,20 +53,31 @@ aliasSccBottomUp procs@(CyclicSCC multi) = do
...
     foldM (\info pspec -> do
         def <- getProcDef pspec
         let (ProcDefPrim _ _ analysis) = procImpln def
-        return $ info ++ [procArgAliasMap analysis]
+        return $ info ++ [extractAliasInfoFromAnalysis analysis]
         ) [] multi
 
 
+extractAliasInfoFromAnalysis :: ProcAnalysis 
+                                -> (AliasMap, AliasMultiSpeczInfo)
```
Comment (pschachte) [inline src/AliasAnalysis.hs:71]: Use AliasInfo type

File: src/AliasAnalysis.hs [hunk 3]
```diff
@@ -51,20 +53,31 @@ aliasSccBottomUp procs@(CyclicSCC multi) = do
...
 currentAliasInfo procs@(CyclicSCC multi) =
     foldM (\info pspec -> do
         def <- getProcDef pspec
         let (ProcDefPrim _ _ analysis) = procImpln def
-        return $ info ++ [procArgAliasMap analysis]
+        return $ info ++ [extractAliasInfoFromAnalysis analysis]
         ) [] multi
 
 
+extractAliasInfoFromAnalysis :: ProcAnalysis
```
Comment (pschachte) [inline src/AliasAnalysis.hs:70]: Document this function

File: test-cases/final-dump/type_list.exp [hunk 4]
```diff
@@ -46,7 +48,9 @@ AFTER EVERYTHING:
 
 
 ++ > public (2 calls)
-0: ++(x#0:type_list.intlist, y#0:type_list.intlist, ?$#0:type_list.intlist): AliasPairs: [($#0,y#0)]
+0: ++(x#0:type_list.intlist, y#0:type_list.intlist, ?$#0:type_list.intlist):
+ AliasPairs: [($#0,y#0)]
```
Comment (pschachte) [inline test-cases/final-dump/type_list.exp:52]: Please check this.  I would have expected $ to be aliased to both x and y.

File: src/AliasAnalysis.hs (hunk 1 above)
Comment (CaviarChen) [reply src/AliasAnalysis.hs]: Added.

File: src/AliasAnalysis.hs (hunk 3 above)
Comment (CaviarChen) [reply src/AliasAnalysis.hs:75]: Added.

File: src/AliasAnalysis.hs (hunk 2 above)
Comment (CaviarChen) [reply src/AliasAnalysis.hs:76]: Actually, `AliasInfo` is `(AliasMap, AliasMultiSpeczInfoLocal)` and here we need `(AliasMap, AliasMultiSpeczInfo)`.

Both `AliasInfo` and `AliasMultiSpeczInfoLocal` are only used in this module. I am not sure if the design is good but my idea is that the local one is used for the analysis (storing temp things and current state) and the non-local one is the actual one that will be stored in the LPVM module (describing the overall info about a procedure).

File: test-cases/final-dump/type_list.exp (hunk 4 above)
Comment (CaviarChen) [reply test-cases/final-dump/type_list.exp:52]: I think the current result is correct. Becasue the whole content from varibale `x` will be reconstructed. The return value only contains the original `y`.

For your reference, here is the code and lpvm instruction:
```
pub def ++(x:intlist, y:intlist):intlist =
    if x = [ ?h | ?t]
    then [h|t++y]
    else y
```

```
++ > public (2 calls)
0: ++(x#0:type_list.intlist, y#0:type_list.intlist, ?$#0:type_list.intlist):
 AliasPairs: [($#0,y#0)]
 AliasMultiSpeczInfo: []
    foreign llvm icmp ne(x#0:type_list.intlist, 0:wybe.int, ?tmp$5#0:wybe.bool)
    case ~tmp$5#0:wybe.bool of
    0:
        foreign llvm move(~y#0:type_list.intlist, ?$#0:type_list.intlist) @type_list:3:5

    1:
        foreign lpvm access(x#0:type_list.intlist, 0:wybe.int, ?h#0:wybe.int)
        foreign lpvm access(~x#0:type_list.intlist, 8:wybe.int, ?t#0:type_list.intlist)
        type_list.++<0>(~t#0:type_list.intlist, ~y#0:type_list.intlist, ?tmp$2#0:type_list.intlist) @type_list:5:13
        foreign lpvm alloc(16:wybe.int, ?tmp$8#0:type_list.intlist)
        foreign lpvm mutate(~tmp$8#0:type_list.intlist, ?tmp$9#0:type_list.intlist, 0:wybe.int, 1:wybe.int, 16:wybe.int, 0:wybe.int, ~h#0:wybe.int)
        foreign lpvm mutate(~tmp$9#0:type_list.intlist, ?$#0:type_list.intlist, 8:wybe.int, 1:wybe.int, 16:wybe.int, 0:wybe.int, ~tmp$2#0:type_list.intlist)
```

File: src/AliasAnalysis.hs (hunk 2 above)
Comment (pschachte) [reply src/AliasAnalysis.hs:76]: I don't think that the fact these things are only used in this module is a good reason not to document them.  At some point, someone is going to need to understand them to maintain the code.  I'm not asking you to document Ting's code (I should have asked her to do that), but please do document the new code and types you're putting in.

File: test-cases/final-dump/type_list.exp (hunk 4 above)
Comment (pschachte) [reply test-cases/final-dump/type_list.exp:52]: Yup, you're right.  I overlooked the fact that it's an int list.

File: src/AliasAnalysis.hs [hunk 5]
```diff
@@ -25,7 +25,10 @@ import           Options       (LogSelection (Analysis))
 import           Util
 
 
--- for local use only
+-- This type is for local use only. It stores interesting
+-- parameters during the analysis. After the analysis, it
+-- will be converted to [AliasMultiSpeczInfo] and stored
+-- in LPVM modure.
```
Comment (pschachte) [inline src/AliasAnalysis.hs:31]: Can you briefly explain what "interesting" means?

File: src/AliasAnalysis.hs (hunk 2 above)
Comment (CaviarChen) [reply src/AliasAnalysis.hs:76]: Sure, I wasn't trying to argue about the documentation in my comment above. I am just saying that the `AliasInfo` is not the same as `(AliasMap, AliasMultiSpeczInfo)`.

**Comment by pschachte** [comment]
Yeah, I didn’t see your change to the doc when I wrote that.  Then I found the change.

I’m still finding github a bit hard to use.

-Peter Schachte

**Pull request merged**